R/clean_data_helpers.R

```diff
@@ -65,12 +38,185 @@ scan_columns <- function(x) {
 #'   data = readRDS(system.file("extdata", "messy_data.RDS",
 #'                              package = "cleanepi"))
 #' )
+#'
+#' @details
+#' For columns of type character, the detected numeric values could actually be
+#' of type Date or date-time. This is because R coerces some Date values into
+#' numeric when the date is imported from an MS Excel file.
+#'
 scan_data <- function(data) {
-  scan_result           <- data.frame(t(apply(data, 2L, scan_columns)))
-  names(scan_result)    <- c("missing", "numeric", "date", "character",
-                             "logical")
-  row_names             <- rownames(scan_result)
-  rownames(scan_result) <- NULL
-  scan_result           <- cbind(Field_names = row_names, scan_result)
+  types       <- vapply(data, typeof, character(1L))
+  scan_result <- vapply(seq_len(ncol(data)), function(col_index) {
+    scan_columns(data[[col_index]], types[[col_index]])
+  }, numeric(7L))
+  scan_result        <- as.data.frame(t(scan_result))
+  names(scan_result) <- c("missing", "numeric", "date", "character",
+                          "logical", "date-time", "factor")
+  scan_result        <- cbind(Field_names = names(data), scan_result)
   return(scan_result)
 }
+
+#' Scan through a double column
+#'
+#' @param x The input vector
+#'
+#' @return A numeric vector with the proportion of the different types of data
+#'    that were detected within the input vector.
+#' @keywords internal
+#'
+scan_in_double <- function(x) {
+  are_factor <- are_date <- are_date_time <- are_character <- are_numeric <-
+    are_logical <- are_na <- 0.0
+  # save the variable length
+  n_rows <- length(x)
+
+  # get the proportion of NA
+  are_na <- round((sum(is.na(x)) / n_rows), 6L)
+  x      <- x[!is.na(x)]
+
+  # doubles are either numeric (attributes = NULL), or Date (has a 'class'
+  # attributes = Date), or date-time (has a 'class' attributes = POSIXt)
+  if ("class" %in% names(attributes(x))) {
+    if ("Date" %in% attributes(x)[["class"]]) {
+      are_date <- round((length(x) / n_rows), 6L)
+    } else if ("POSIXt" %in% attributes(x)[["class"]]) {
+      are_date_time <- round((length(x) / n_rows), 6L)
+    }
```

Comment:
As mentioned, this is not the proper way to test class in R. Use `inherits()`. Please don't use complex workarounds to go around this.